bindings/python/pyiree/jax/frontend.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+
+import pyiree.compiler2 as compiler
+import pyiree.compiler2.xla
+from pyiree import rt
+
+try:
+  import jax
+except ModuleNotFoundError as e:
+  raise ModuleNotFoundError("pyiree.jax requires 'jax' and 'jaxlib' to be "
+                            "installed in your python environment.") from e
+
+# pytype thinks pyiree.jax is jax.
+# pytype: disable=module-attr
+
+__all__ = [
+    "aot",
+    "is_available",
+    "jit",
+]
+
+_BACKENDS = ["vmla", "llvmaot", "vulkan"]
+_BACKEND_TO_DRIVER = {
+    "vmla": "vmla",
+    "llvmaot": "dylib",
+    "vulkan": "vulkan",
+}
+_BACKEND_TO_TARGETS = {
+    "vmla": ("vmla",),
+    "llvmaot": ("dylib-llvm-aot",),
+    "vulkan": ("vulkan-*",),
+}
+
+
+def is_available():
+  """Determine if the IREE–XLA compiler are available for JAX."""
+  return compiler.xla.is_available()
+
+
+def aot(function, *args, **options):
+  """Traces and compiles a function, flattening the input args.
+
+  This is intended to be a lower-level interface for compiling a JAX function to
+  IREE without setting up the runtime bindings to use it within Python. A common
+  usecase for this is compiling to Android (and similar targets).
+
+  Args:
+    function: The function to compile.
+    args: The inputs to trace and compile the function for.
+    **kwargs: Keyword args corresponding to xla.ImportOptions or CompilerOptions
+  """
+  xla_comp = jax.xla_computation(function)(*args)
+  hlo_proto = xla_comp.as_serialized_hlo_module_proto()
+  binary = compiler.xla.compile_str(hlo_proto, **options)
+  return binary
+
+
+# A more JAX-native approach to jitting would be desireable here, however
+# implementing that reasonably would require using JAX internals, particularly
+# jax.linear_util.WrappedFun and helpers. The following is sufficient for many
+# usecases for the time being however.
+
+
+class _JittedFunction:
+
+  def __init__(self, function, driver: str, **options):
+    self._function = function
+    self._driver_config = rt.Config(driver)
+    self._options = options
+    self._memoized_signatures = {}
+
+  def _get_signature(self, args_flat):
+    args_flat = [rt.normalize_value(arg) for arg in args_flat]
+    return tuple([(arg.shape, arg.dtype) for arg in args_flat])
```

Comment:
You can avoid materializing the list just to cast to a tuple by using a naked generator expression:
  return tuple((arg.shape, arg.dtype) for arg in args_flat)